[4.0] [NO CACHE] Web Services

wilsonge opened the pull request:
This is a base for a web services implementation for Joomla 4

## What this has got in
### Core
- Content Negotiation (core provides a basic api integration using the https://github.com/tobscure/json-api library)
- Plugin based routes to allow custom endpoints to be added
- Authentication on endpoints using users basic auth (through plugins to allow future oAuth integrations)
- Exceptions added to allow quality error handling

### Integrations/Testing
- Sample com_content integration
- Basic system tests

## What does this not have in
- oAuth authentication
- Ability to turn off web services - although there is an auth layer in front of it
- Some sort of fundamental change to our error handling (the way we set up errors in JTable and JModel mean we'll likely never be able to have nice error handling with the current model system. In many places we're likely going to need to add 500's just because we won't be able to detect the right exceptions. In the future we should consider a new mvc layer based on the new entity layer https://github.com/joomla-framework/entities but this will take longer and be much harder to implement. The fundamental base application would not be expected to change but extension MVC's will have to to have a nice working integration.

## Thanks
- @mbabker for large amounts of consulting and framework contributions especially around the router as well as for the code for api part of the ComponentInstaller Adapter
- @muhakh and @cokencorn for their work on their GSOC project which formed a large amount of the base application 
- @isacandrei for work bug testing and working on various next gen parts

## Before moaning about a lack of feature X
Please read the specification https://joomla-projects.github.io/gsoc18_webservices/ - this has long been agreed on and is by no means perfect. Think about if it's absolutely essential for a day 1 integration or not. What we have in this PR is already significantly more than wordpress has and is probably equal to what's available in the other top level CMS'

## Testing
- When applying the patch either do a fresh install of Joomla or discover install the new plugins
- In the specification docs above there's a "installing the latest codebase" menu item which contains sample API calls you can make (currently this is just for com_content)
- Please ensure with all the library changes that nothing has broken in the backend/frontend too

wilsonge committed:
Redraft of webservices

dgrammatiko commented:
Merge it!

brianteeman commented on libraries/src/Access/Exception/AuthenticationFailed.php:
```diff
@@ -0,0 +1,20 @@
+defined('JPATH_PLATFORM') or die;
+
+/**
+ * Exception class defining a authentication failed event
```
==> an not a

wilsonge replied:
Thanks :) Fixed

brianteeman commented on libraries/src/Access/Exception/AuthenticationFailed.php:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Exception class defining a authentication failed event
+ *
+ * @since  3.6.3
```
3.6.3 ??

wilsonge replied:
Thanks :) Fixed

brianteeman commented:
Sorry to be pedantic but the correct term according to the [w3c](https://www.w3.org/2002/ws/) it is "Web Services" not "webservices"

As this is new to Joomla its best to get it right from the beginning

wilsonge commented:
@brianteeman I've amended the language strings and the doc blocks where we refer to web services as a noun. I've left the plugin type as it is - because obviously it can't have spaces and i've found from practical experience with editors-xtd that most people struggle with the concept of removing the `-` from the plugin class names - so it's more practical to keep the group as a single word I *think*. Happy to be challenged on that if that's the majority tho

dgrammatiko commented on api/index.php:
```diff
@@ -0,0 +1,36 @@
+		str_replace(
+			array('{{PHP_VERSION}}', '{{BASEPATH}}'),
+			array(JOOMLA_MINIMUM_PHP, 'http://' . $_SERVER['SERVER_NAME'] . '/'),
+			file_get_contents(dirname(__FILE__) . '/../templates/system/incompatible.html')
```
This won’t fly...
Either you always return JSON or this is already a broken API

wilsonge replied:
Fixed in a somewhat dirty way

dgrammatiko replied:
Nice. It's the same "dirty way" we're using for the HTML part as well... 😂

mbabker commented on api/includes/framework.php:
```diff
@@ -0,0 +1,88 @@
+{
+	if (file_exists(JPATH_INSTALLATION . '/index.php'))
+	{
+		header('Location: ../installation/index.php');
```
Don't do a redirect here

mbabker commented on api/includes/framework.php:
```diff
@@ -0,0 +1,88 @@
+	}
+	else
+	{
+		echo 'No configuration file found and no installation code available. Exiting...';
```
Make JSON

mbabker commented:
App needs to be registered in `Joomla\CMS\Application\ApplicationHelper::getClientInfo()`

(Side note, this method should be redone to use the event system instead of hardcoding a list...)

bembelimen commented on api/components/com_content/View/Articles/JsonapiView.php:
```diff
@@ -0,0 +1,30 @@
+	 * @var  string
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected $fieldsToRender = ['id', 'typeAlias', 'asset_id', 'title', 'introtext', 'state', 'catid', 'created'];
```
Don't trust ```state``` :)

wilsonge replied:
Why?

Hackwar commented on api/components/com_content/Controller/ArticleController.php:
```diff
@@ -0,0 +1,30 @@
+	 * @var    string
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected $contentType = 'articles';
```
I'm rubbing myself a bit on this being plural in a controller that is singular. We are a bit inconsistent here elsewhere, too...

wilsonge replied:
Made the controller plural

Hackwar commented on libraries/src/Application/ApiApplication.php:
```diff
@@ -0,0 +1,338 @@
+
+		/**
+		 * Now we have an API perform content negotation to ensure we have a valid header. Assume if the route doesn't
+		 * tell us otherwise it uses the pain JSON API
```
`it uses the pLain JSON API`

Hackwar commented on libraries/src/Application/ApplicationHelper.php:
```diff
@@ -144,6 +144,12 @@ public static function getClientInfo($id = null, $byName = false)
 			$obj->path = JPATH_INSTALLATION;
 			self::$_clients[2] = clone $obj;
+
+			// Installation Client
```
`// API Client`

Hackwar commented on libraries/src/Document/JsonapiDocument.php:
```diff
@@ -0,0 +1,201 @@
+use Tobscure\JsonApi\ElementInterface;
+
+/**
+ * JDocumentJsonapi class, provides an easy interface to parse output in JSON-API format.
```
`JsonapiDocument`

Hackwar commented:
I like this, but we would indeed need at least a way to switch it off. Maybe simply adding a switch in configuration.php?

wilsonge commented:
@Hackwar addressed all your comments. So I deliberately didn't put in the switch yet because I want it in a separate PR. Reason being - there is an argument that inter-component webservices should always be available. The switch should be a limit for the public facing webservices (I honestly can see this both ways). As a result I want to deliberately NOT have that discussion in this PR and leave it for later and instead introduce the base application here and let that discussion be done in a separate follow up PR (which I commit to doing in order to allow that conversation to take place)

mbabker commented:
The API app should behave the same as the admin app as far as folder structure goes (since the requests are going to `/api/index.php` and not through the root `/index.php`).

sanderpotjer commented:
First off all: thank you @wilsonge and all others worked on this Web Services PR, great work! 

We have been looking into this PR this afternoon. In the past we have developed a custom REST API solution to serve Joomla articles for our customers. So as a test-case we are now doing this again, but by using this new core feature (and possibly a custom web services plugin). 

A couple comments so far:

**1. Documentation**
You asked us to read https://joomla-projects.github.io/gsoc18_webservices/, which kind of pushed us into the incorrect direction for testing this PR ;-) The example URLs on https://joomla-projects.github.io/gsoc18_webservices/?specification/chapters/urls-and-routing.md are no longer valid. It might be good to update that page, or have some up to date documentation for people that want to test this PR. And mention that the authentication works via the Joomla user accounts. 

**2. Requests not working**
We confirm the "not working" items mentioned by @manojLondhe

**3. SEF support**
The GET articles list is working, but it seems that as soon as we turn "Use URL Rewriting" on, or completely disable the "Search Engine Friendly URLs" the API will return a "Resource not found" error response.

**4. ACL**
Currently, the only authentication to access the API is done by checking the password of a user. We understand that additional API authentication plugins can be created, but would it not be good to add an ACL check as well? We do see a reference to an action "core.login.api" in https://github.com/joomla/joomla-cms/pull/23424/files#diff-45ae31282d7eb57ef5d8dc26df256d9fR286 but that is not implemented further in the web services / Joomla permission interface. We do think it would be a good idea to add that "core.login.api" action and check for the permission. In this way you don't grant anyone with an account on the website access to the API by default. There are situations that only a specific "API users" should be able to access the API. The "core.login.api" is also in line with the offline, admin and site login actions. As a bonus: this might be the solution as well to "turn off web services" by simply not allowing this action for the Public user group.

These are some first comments based on our first impressions. Its great to see Web Services are on the horizon of Joomla 4 and will try to test/contribute as much as possible to make this happen for Joomla 4 :)

wilsonge commented:
> Documentation

Sorry forgot those existed. This had the 'correct' URLs on https://github.com/joomla-projects/gsoc18_webservices/blob/master/manual/en-US/about-joomla/installing-the-repo.md but either way I've normalised everything (I've now also expanded the API Auth section). Note since your test I've added in the version prefix. Hopefully this now reflects what's in the product (once github propagates the commits into the docs)

There's a possibility I might swap the Basic Auth out for a token based system before this goes stable. But for now at least it will be basic auth.

> Requests not working

Did you guys clear the libraries/autoload_psr4.php file to autoload the components api sections? Because if you didn't that would definitely cause the issue you're seeing. I have also fixed an issue with the edit and post commands. Looks like I still have a bug in the delete one which I'll look into tomorrow.

> SEF support
I thought I'd covered this with https://github.com/wilsonge/joomla-cms/blob/webservices-version2/libraries/src/Router/ApiRouter.php#L117-L128 - might need some help to test this if it's not working as I'm mostly working local and haven't set up the whole htaccess stuff.

> ACL

Oops section missing from com_config now added. Hopefully that's the permission you were looking for (note by default I'd picked admin users only but again that can all be negotiated)

wilsonge commented:
Delete button fixed. That should be all the endpoints working again

yvesh commented:
I have tested this item :white_check_mark: successfully on bd17e194da29a3dbb091ff8a9abf5d8905eaa734<br /><br />Basic API calls work as expected.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23424">issues.joomla.org/tracker/joomla-cms/23424</a>.</sub>

wilsonge commented:
@SniperSister @zero-24 can you guys have a look at RIPS please (carefully) and if it's good i'm merging this

alikon commented on administrator/components/com_admin/sql/updates/postgresql/4.0.0-2019-01-05.sql:
```diff
@@ -0,0 +1,3 @@
+INSERT INTO "#__extensions" ("extension_id", "name", "type", "element", "folder", "client_id", "enabled", "access", "protected", "manifest_cache", "params", "custom_data", "system_data", "checked_out", "checked_out_time", "ordering", "state") VALUES
+(493, 0, 'plg_api-authentication_basic', 'plugin', 'basic', 'api-authentication', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
```
```suggestion
(493, 0, 'plg_api-authentication_basic', 'plugin', 'basic', 'api-authentication', 0, 1, 1, 0, '', '{}', 0, '1970-01-01 00:00:00', 0, 0),
```

alikon replied:
default for postgresql

alikon commented on administrator/components/com_admin/sql/updates/postgresql/4.0.0-2019-01-05.sql:
```diff
@@ -0,0 +1,3 @@
+INSERT INTO "#__extensions" ("extension_id", "name", "type", "element", "folder", "client_id", "enabled", "access", "protected", "manifest_cache", "params", "custom_data", "system_data", "checked_out", "checked_out_time", "ordering", "state") VALUES
+(493, 0, 'plg_api-authentication_basic', 'plugin', 'basic', 'api-authentication', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
+(494, 0, 'plg_webservices_content', 'plugin', 'content', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
```
```suggestion
(494, 0, 'plg_webservices_content', 'plugin', 'content', 'webservices', 0, 1, 1, 0, '', '{}', 0, '1970-01-01 00:00:00', 0, 0),
```

alikon replied:
default for postgresql

alikon commented on installation/sql/postgresql/joomla.sql:
```diff
@@ -685,6 +685,8 @@ INSERT INTO "#__extensions" ("extension_id", "package_id", "name", "type", "elem
 (491, 0, 'plg_installer_override', 'plugin', 'override', 'installer', 0, 1, 1, 1, '', '', 0, '1970-01-01 00:00:00', 4, 0),
 (492, 0, 'plg_quickicon_overridecheck', 'plugin', 'overridecheck', 'quickicon', 0, 1, 1, 1, '', '', 0, '1970-01-01 00:00:00', 0, 0),
 (493, 0, 'plg_extension_finder', 'plugin', 'finder', 'extension', 0, 1, 1, 0, '', '', 0, '1970-01-01 00:00:00', 0, 0),
+(494, 0, 'plg_api-authentication_basic', 'plugin', 'basic', 'api-authentication', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
```
```suggestion
(494, 0, 'plg_api-authentication_basic', 'plugin', 'basic', 'api-authentication', 0, 1, 1, 0, '', '{}', 0, '1970-01-01 00:00:00', 0, 0),
```

alikon commented on installation/sql/postgresql/joomla.sql:
```diff
@@ -685,6 +685,8 @@ INSERT INTO "#__extensions" ("extension_id", "package_id", "name", "type", "elem
 (492, 0, 'plg_quickicon_overridecheck', 'plugin', 'overridecheck', 'quickicon', 0, 1, 1, 1, '', '', 0, '1970-01-01 00:00:00', 0, 0),
 (493, 0, 'plg_extension_finder', 'plugin', 'finder', 'extension', 0, 1, 1, 0, '', '', 0, '1970-01-01 00:00:00', 0, 0),
+(494, 0, 'plg_api-authentication_basic', 'plugin', 'basic', 'api-authentication', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
+(495, 0, 'plg_webservices_content', 'plugin', 'content', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
```
```suggestion
(495, 0, 'plg_webservices_content', 'plugin', 'content', 'webservices', 0, 1, 1, 0, '', '{}', 0, '1970-01-01 00:00:00', 0, 0),
```

wilsonge commented:
Thanks @alikon

wilsonge commented:
@SniperSister @zero-24 looks like still some RIPS issues - although most have been solved elsewhere

zero-24 commented:
Rips has been cancelled just restarted it here: http://ci.joomla.org/joomla/joomla-cms/15490

wilsonge commented:
@zero-24 Sorry my bad I started playing around with the API tests. http://ci.joomla.org/joomla/joomla-cms/15495/19 is a fixed build

Oracions2410 commented:
Hello everyone .. I am interested in this project and wish to participate. Can you guide me please?

Oracions2410 commented:
I have exceptional skills in PHP. And a good dose of motivation. Help me please !

alikon commented:
@Oracions2410 did you mean the gsoc 2019 project https://docs.joomla.org/GSOC_2019_Project_Ideas#Project_I:_Webservices_in_Joomla ?

wilsonge commented:
Merging. Thanks again to all those who gave input, and the gsoc students who worked on this project. It's hugely appreciated!

Oracions2410 commented:
Yes. I just discovered the world of open source and I think Gsoc is a good start!

wilsonge commented:
@Oracions2410 that's awesome. If you drop one of the GSOC admins on that page a email they'll drop you in some of our chat channels where we can discuss details of the project etc :)

Oracions2410 commented:
@wilsonge 
Excuse me, on the project you are mentioned as a main mentor. Can you give me the chanels where I can find help please

Sandra97 commented:
@Oracions2410 So that we can add you to our communication channel, please drop us your email on our mailing list https://groups.google.com/forum/#!forum/jgsoc

anibalsanchez commented:
Preparing the talk for the [Mastermind Podcast](https://mastermindjoomla.com/), these are my notes about Web Services quick testing 101:

- Step 1: Download and Install **Joomla 4 Alpha 7**
- Step 2: Notice that there two new plugin groups: **api-authentication** and **webservices**
- Step 3: Notice that there two new plugins: **API Authentication - Basic Auth** and **Web Services - Content**
- Step 4: J4 A7 has API Authentication based on **Http Basic Authentication**. So, yes, you have to configure .htpasswd to test the API. Ref: http://httpd.apache.org/docs/2.4/programs/htpasswd.html
- Step 5: Of course, if you are a little bit lazy, you can go straight to _plugins/api-authentication/basic/basic.php_, and change the authentication or create your own plugin.
- Step 6: Create an article and access the REST API for the Content component. For instance: http://mytest.com/api/index.php/v1/article/1

For more information:

- Joomla! Webservices Specification - https://joomla-projects.github.io/gsoc18_webservices/

wilsonge commented:
@Oracions2410 I am - but i was preparing for a presentation i was giving in paris so I wasn't really available. I'm now back up and running. either way however you needed to be added into glip :)

wilsonge commented:
@anibalsanchez i don't think you need to configure .htpasswd to test the api?

anibalsanchez commented:
In my tests, it's been giving me Forbidden errors until I configured the auth.

Carlos has a similar experience.

carcam commented:
Thanks everyone for this awesome feature!!

I can confirm the server doesn't ask me for Username and Password as it's supposed to. I'm using a Lando box using Anibal Lando Recipe for Joomla 4 so I guess this has to do with the server configuration as stated in some comments I found on the internet about this auth method not working out of the box in CGI servers.

In any case I would point out just 2 issues:
1. with some server configurations you cannot authenticate out of the box and also 
2. Probably checking if user has an account for public content should avoided. I understand the idea of this is to be a "content leak stopper" but we can do better ;)

In any case, again, I was really excited with having this feature when you announced it and I event made a whole episode of my podcast about it because it's awesome to keep creating great solutions for our community and our clients.

wilsonge commented:
> I can confirm the server doesn't ask me for Username and Password as it's supposed to. I'm using a Lando box using Anibal Lando Recipe for Joomla 4 so I guess this has to do with the server configuration as stated in some comments I found on the internet about this auth method not working out of the box in CGI servers.

Wait are you making requests in the browser or something? Nothing is supposed to prompt you for a password?

anibalsanchez commented:
I already told him that he has to configure the .htpassword file to configure the Apache Http Basic Authentication.

I agree with Carlos that if the plugin is enabled, then public content (articles) must be allowed to the Guest user.
